### DISCUSSION TITLE
How to implement scalar/aggregate functions with their own config/parameter?

### DISCUSSION THREAD

**Discussion by Yohahaha** [facebookincubator/velox#4611] [Ideas]
### Background Info

When add `first/last` Spark aggregate function, @mbasmanova points out there are some inconsistent behaviors for scalar function and aggregate function to provide configureable implementation, see https://github.com/facebookincubator/velox/pull/4578#discussion_r1166202190

Spark [size](https://spark.apache.org/docs/3.3.1/api/sql/#size) function has `spark.sql.legacy.sizeOfNull` config to control its behavior, and Velox provide a QueryConfig for it https://facebookincubator.github.io/velox/configs.html#spark-legacy-size-of-null. 
https://github.com/facebookincubator/velox/blob/e1305d4c4634de28aa03e24c5ca07de7b98f337c/velox/functions/sparksql/Size.cpp#L30-L35

Spark first aggregate function provide a ignoreNull parameter in SQL usage: `select first(expr, ignoreNull) from tmp`. However, Velox aggregate function does not have any config class to initialize itself compared with scalar function.

#### For function parameters in SQL
Parameters will be parsed by parser as each function's own member, does not shared in session or query. These functions should provide different implementations to fulfill it, for example, `select first(expr) as not_ignore_null_expr, first(expr, true) as ignore_null_expr from tmp`, downstream engine which use Velox for accelerating should invoke first() and first_ignore_null() separately.

#### For function depends on config
Some functions depends on static/session config to exhibit different behavior, we should pass a config class in its initialize method. For example, Spark `size` function depends on spark.sql.legacy.sizeOfNull, Presto `date_add` function depends on session timezone.

However, we lack of the ability for pass session config to aggregate functions. Consider if someone wants add a new aggregate function depends on session timezone, how could we implement it?

CC @mbasmanova

**Comment by mbasmanova** [comment]
> Parameters will be parsed by parser as each function's own member, does not shared in session or query.

@Yohahaha I didn't know Spark allows this. Curious if you could share some pointer to documentation and example queries. Assuming ignore-nulls is such parameter, I agree with you that Velox needs to provide 2 separate function implementations.

**Reply by Yohahaha** [reply]
Spark last aggregate function's source code may be helpful. https://github.com/apache/spark/blob/ca2ddf3c2079dda93053e64070ebda1610aa1968/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Last.scala#L53

Example query and output
```
spark-sql> SELECT last(col), last(col, true) FROM VALUES (10), (5), (NULL) AS tab(col);
NULL	5
Time taken: 0.272 seconds, Fetched 1 row(s)
```

**Reply by mbasmanova** [reply]
@Yohahaha Thank you for the example. I'm curious how last(col, true) is planned. You mentioned that 'true' is not a column that is being passed into the aggregate. Would you share a query plan?

**Reply by Yohahaha** [reply]
@mbasmanova 
```
spark-sql> explain extended SELECT last(col), last(col, true) FROM VALUES (10), (5), (NULL) AS tab(col);
== Parsed Logical Plan ==
'Project [unresolvedalias(last('col, false), None), unresolvedalias('last('col, true), None)]
+- 'SubqueryAlias tab
   +- 'UnresolvedInlineTable [col], [[10], [5], [null]]

== Analyzed Logical Plan ==
last(col): int, last(col): int
Aggregate [last(col#24, false) AS last(col)#26, last(col#24, true) AS last(col)#27]
+- SubqueryAlias tab
   +- LocalRelation [col#24]

== Optimized Logical Plan ==
Aggregate [last(col#24, false) AS last(col)#26, last(col#24, true) AS last(col)#27]
+- LocalRelation [col#24]

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- HashAggregate(keys=[], functions=[last(col#24, false), last(col#24, true)], output=[last(col)#26, last(col)#27])
   +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=33]
      +- HashAggregate(keys=[], functions=[partial_last(col#24, false), partial_last(col#24, true)], output=[last#32, valueSet#33, last#34, valueSet#35])
         +- LocalTableScan [col#24]
```

**Reply by mbasmanova** [reply]
@Yohahaha This plan suggests that 'last' function is called with 2 inputs: one is a column, another a constant. Is this the case? Should we then allow last to be called with 2 arguments in Velox? We can mark second argument as constant in the signature.

**Reply by Yohahaha** [reply]
> Should we then allow last to be called with 2 arguments in Velox? We can mark second argument as constant in the signature.

Both ways ok to me. If second constant argument is not a boolean, we should declared in signature.

**Reply by mbasmanova** [reply]
Sounds like a cleaner solution is to define 2 versions of last function: last(x) and last(x, ignoreNull). Let's do that.

**Reply by Yohahaha** [reply]
I remember why I choose last and last_ignore_null instead of last(x, ignoreNull).

If we use last(x, ignoreNull) as signature, it required input row vectors must have two column, and second column vector should be constant encoding. But consider above query plan, Scan operator just return row vectors contains only one column, how do we know the value of ignoreNull? Also in aggregate fuzzer, if the signature is last(x, boolean), we can not generate the row vectors contains only one column.
I think current aggregate function signature mechanism in Velox does not allow to define an argument which not related with row vectors,  can we?

**Reply by Yohahaha** [reply]
For example, Gluten use substrait to communicate with Velox, when we parse last(x, ignoreNull) from substrait to Velox, we can choose last or last_ignore_null depends on ignoreNull value, but we have no way to set ignoreNull as Velox aggregate function's argument.

**Reply by mbasmanova** [reply]
This should be possible. When creating CallTypedExpr for the last(x, true), you can pass ConstantTypedExpr as the second parameter.

**Reply by Yohahaha** [reply]
I got your points, and find related Issue #3993 #4140, I will follow up with first/last aggregate function refactor, thanks!

**Reply by Yohahaha** [reply]
> This plan suggests that 'last' function is called with 2 inputs: one is a column, another a constant. Is this the case?

Actually, `last` function does not accept 2 inputs, above plan is just a toString method's output for readability. `last` only accept 1 input `child`, and second argument in its constructor is not Expression, just a boolean variable. Spark aggregate function use `updateExpressions` to update group value like `addRawInput/addSingleGroupRawInput`, last aggregate functions's `updateExpressions` only compute `child` expression and use `ignoreNull` to decide if branch.
https://github.com/apache/spark/blob/87ccfc2402394d168a90c8df494482c7a1d6a552/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Last.scala#L90-L102

As for `sort_array` in #3993, its constructor has 2 arguments, base and ascendingOrder, are all Expression, so it could be called with 2 inputs.
https://github.com/apache/spark/blob/9e17731f46aa26348e69f42d1a96882186022b80/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L1032-L1047

So, I think for first/last like aggregate functions, Velox provides last(x) and last_ignore_null(x) are make sense.

**Reply by mbasmanova** [reply]
@Yohahaha Thank you for clarifying. Makes sense to me. Appreciate the detailed explanation. Thank you for digging deep.

**Comment by mbasmanova** [comment]
> Some functions depends on static/session config to exhibit different behavior, we should pass a config class in its initialize method. 

Makes sense to me. Do we have an example of an aggregate function that needs access to session config?

**Reply by Yohahaha** [reply]
I don't find such functions either, maybe I'm wrong.

**Comment by Yohahaha** [comment]
I see Presto has similar `ignoreNull` function, how do we handle these? https://prestodb.io/docs/current/functions/window.html#value-functions

**Reply by mbasmanova** [reply]
@Yohahaha This is part of WindowNode

```
class WindowNode : public PlanNode {
 public:
  ...

  struct Function {
    CallTypedExprPtr functionCall;
    Frame frame;
    bool ignoreNulls;

    folly::dynamic serialize() const;

    static Function deserialize(const folly::dynamic& obj);
  };

  /// @param windowColumnNames specifies the output column
  /// names for each window function column. So
  /// windowColumnNames.length() = windowFunctions.length().
  WindowNode(
      PlanNodeId id,
      std::vector<FieldAccessTypedExprPtr> partitionKeys,
      std::vector<FieldAccessTypedExprPtr> sortingKeys,
      std::vector<SortOrder> sortingOrders,
      std::vector<std::string> windowColumnNames,
      std::vector<Function> windowFunctions,
      PlanNodePtr source);
```

@aditi-pandit Aditi, would you confirm whether ignoreNulls flag is currently being respected or not?

**Reply by aditi-pandit** [reply]
@mbasmanova : ignoreNulls flag is not respected yet. This flag is only for value functions (first, last, nth, lead, lag) for windows. We have some PRs in flight for adding this behavior since we have got to lead/lag functions now.

**Reply by mbasmanova** [reply]
@aditi-pandit Aditi, thank you for clarifying. It would be nice to add this functionality.

CC: @zacw7 Zac, FYI, this a gap currently present in the implementation of window functions.
